src/Application/UI/Presenter.php

```diff
@@ -779,87 +779,23 @@ protected function createRequest($component, $destination, array $args, $mode)
-		$a = strpos($destination, '#');
-		if ($a === FALSE) {
-			$fragment = '';
-		} else {
-			$fragment = substr($destination, $a);
-			$destination = substr($destination, 0, $a);
-		}
-
-		// 2) ?query syntax
-		$a = strpos($destination, '?');
-		if ($a !== FALSE) {
-			parse_str(substr($destination, $a + 1), $args);
-			$destination = substr($destination, 0, $a);
-		}
-
-		// 3) URL scheme
-		$a = strpos($destination, '//');
-		if ($a === FALSE) {
-			$scheme = FALSE;
-		} else {
-			$scheme = substr($destination, 0, $a);
-			$destination = substr($destination, $a + 2);
-		}
-
-		// 4) signal or empty
-		if (!$component instanceof self || substr($destination, -1) === '!') {
-			$signal = rtrim($destination, '!');
-			$a = strrpos($signal, ':');
-			if ($a !== FALSE) {
-				$component = $component->getComponent(strtr(substr($signal, 0, $a), ':', '-'));
-				$signal = (string) substr($signal, $a + 1);
-			}
-			if ($signal == NULL) {  // intentionally ==
-				throw new InvalidLinkException('Signal must be non-empty string.');
-			}
-			$destination = 'this';
-		}
-
-		if ($destination == NULL) {  // intentionally ==
-			throw new InvalidLinkException('Destination must be non-empty string.');
-		}
-
-		// 5) presenter: action
-		$current = FALSE;
-		$a = strrpos($destination, ':');
-		if ($a === FALSE) {
-			$action = $destination === 'this' ? $this->action : $destination;
-			$presenter = $this->getName();
-			$presenterClass = get_class($this);
-
-		} else {
-			$action = (string) substr($destination, $a + 1);
-			if ($destination[0] === ':') { // absolute
-				if ($a < 2) {
-					throw new InvalidLinkException("Missing presenter name in '$destination'.");
-				}
-				$presenter = substr($destination, 1, $a - 1);
-
-			} else { // relative
-				$presenter = $this->getName();
-				$b = strrpos($presenter, ':');
-				if ($b === FALSE) { // no module
-					$presenter = substr($destination, 0, $a);
-				} else { // with module
-					$presenter = substr($presenter, 0, $b + 1) . substr($destination, 0, $a);
-				}
-			}
-			if (!$this->presenterFactory) {
-				throw new Nette\InvalidStateException('Unable to create link to other presenter, service PresenterFactory has not been set.');
-			}
-			try {
-				$presenterClass = $this->presenterFactory->getPresenterClass($presenter);
-			} catch (Application\InvalidPresenterException $e) {
-				throw new InvalidLinkException($e->getMessage(), NULL, $e);
-			}
+		$parsedLinkDestination = $this->parseLinkDestination($component, $destination);
+		$destination = $parsedLinkDestination['destination'];
+		$scheme = $parsedLinkDestination['scheme'];
+		$presenter = $parsedLinkDestination['presenter'];
+		$presenterClass = $parsedLinkDestination['presenterClass'];
+		$action = $parsedLinkDestination['action'];
+		$current = $parsedLinkDestination['current'];
+		$component = $parsedLinkDestination['callee'];
+		$signal = $parsedLinkDestination['signal'];
+		$fragment = $parsedLinkDestination['fragment'];
```

Comment:
That's exactly what I didn't want to do.

Ondřej Mirtes




On Tue, Mar 29, 2016 at 2:35 AM -0700, "Jan Tvrdík" <notifications@github.com> wrote:












In src/Application/UI/Presenter.php:
> -			}
> -			try {
> -				$presenterClass = $this->presenterFactory->getPresenterClass($presenter);
> -			} catch (Application\InvalidPresenterException $e) {
> -				throw new InvalidLinkException($e->getMessage(), NULL, $e);
> -			}
> +		$parsedLinkDestination = $this->parseLinkDestination($component, $destination);
> +		$destination = $parsedLinkDestination['destination'];
> +		$scheme = $parsedLinkDestination['scheme'];
> +		$presenter = $parsedLinkDestination['presenter'];
> +		$presenterClass = $parsedLinkDestination['presenterClass'];
> +		$action = $parsedLinkDestination['action'];
> +		$current = $parsedLinkDestination['current'];
> +		$component = $parsedLinkDestination['callee'];
> +		$signal = $parsedLinkDestination['signal'];
> +		$fragment = $parsedLinkDestination['fragment'];



You can use extract



—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub